File: lib/nibble/frontend/nibble-admin/conditions/__tests__/conditions.test.ts
import { afterEach, describe, expect, it } from 'vitest'
import cases from '../../../../../../test/fixtures/files/nibble_conditions.json'
import { isVisible, registerCondition, unregisterCondition } from '..'

type Case = {
  description: string
  config: Record<string, unknown>
  values: Record<string, unknown>
  root_values?: Record<string, unknown>
  path?: string
  prefix?: string
  visible: boolean
}

describe('isVisible (shared with Nibble::Conditions)', () => {
  for (const example of cases as Case[]) {
    it(example.description, () => {
      const visible = isVisible(example.config, {
        values: example.values,
        rootValues: example.root_values,
        path: example.path ?? null,
        prefix: example.prefix ?? null,
      })
      expect(visible).toBe(example.visible)
    })
  }
})

describe('custom conditions', () => {
  afterEach(() => {
    unregisterCondition('longer_than')
    unregisterCondition('never')
  })

  it('receive params and the target value', () => {
    registerCondition('longer_than', ({ params, target }) => String(target).length > Number(params[0]))
    expect(isVisible({ if: { title: 'custom longer_than:3' } }, { values: { title: 'Hello' } })).toBe(true)
    expect(isVisible({ if: { title: 'custom longer_than:10' } }, { values: { title: 'Hello' } })).toBe(false)
  })

  it('are inverted by unless even without a target, like the server', () => {
    registerCondition('never', () => false)
    expect(isVisible({ unless: 'custom never' }, { values: {} })).toBe(true)
  })

  it('fail loud when unregistered', () => {
    expect(() => isVisible({ if: 'custom missing' }, { values: {} })).toThrow("'missing' isn't registered")
  })
})

File: test/lib/nibble/conditions_test.rb
require "test_helper"

# Shared with the CP evaluator's tests (vendor/nibble/frontend/nibble-admin/conditions) so server and editor agree on visibility.
class Nibble::ConditionsTest < ActiveSupport::TestCase
  CASES = JSON.parse(Rails.root.join("test/fixtures/files/nibble_conditions.json").read)

  teardown { Nibble::Conditions.reset! }

  CASES.each do |example|
    test "shared case: #{example['description']}" do
      visible = Nibble::Conditions.visible?(example["config"], values: example["values"], root_values: example["root_values"],
        path: example["path"], prefix: example["prefix"])
      assert_equal example["visible"], visible
    end
  end

  test "custom conditions receive params and target, and unregistered ones fail loud" do
    Nibble::Conditions.register("longer_than") { |params:, target:, **| target.to_s.length > params.first.to_i }

    assert Nibble::Conditions.visible?({ "if" => { "title" => "custom longer_than:3" } }, values: { "title" => "Hello" })
    assert_not Nibble::Conditions.visible?({ "if" => { "title" => "custom longer_than:10" } }, values: { "title" => "Hello" })

    Nibble::Conditions.register("never") { |**| false }
    assert Nibble::Conditions.visible?({ "unless" => "custom never" }, values: {}), "a target-less custom condition is inverted by unless"

    assert_raises(Nibble::Error, match: /'missing' isn't registered/) { Nibble::Conditions.visible?({ "if" => "custom missing" }, values: {}) }
  end
end
